Fix `go` directive in `go.mod`

matthewhughes934 opened the pull request:
The docs[1] say:

> The version must be a valid Go version, such as 1.9, 1.14, or 1.21rc1.

A 'Go version' is a go release[2], but there's no release of Go `1.22` (compare with Go `1.20` that _does_ have a release) the first release of Go `1.22` was Go `1.22.0` so use that here.

This will cause issue in any environment where parsing this module triggers a toolchain download (i.e. something running Go `1.21`, since before then the directive was only advisory, and with the `GOTOOLCHAIN` env var not set to `none`), for example:

    $ docker run --rm --interactive --volume "$PWD:/work" --workdir /work golang:1.21.0 go mod tidy
    go: downloading go1.22 (linux/amd64)
    go: download go1.22 for linux/amd64: toolchain not available

This change should have no impact on any setup that works with the current directive.

There's more detail and discussions around this at[4]

[1] https://go.dev/ref/mod#go-mod-file-go
[2] https://go.dev/doc/toolchain#version
[3] https://go.dev/doc/devel/release
[4] https://github.com/golang/go/issues/62278

<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

matthewhughes934 committed:
Fix `go` directive in `go.mod`

The docs[1] say:

> The version must be a valid Go version, such as 1.9, 1.14, or 1.21rc1.

A 'Go version' is a go release[2], but there's no release of Go `1.22`
(compare with Go `1.20` that _does_ have a release) the first release of
Go `1.22` was Go `1.22.0` so use that here.

This will cause issue in any environment where parsing this module
triggers a toolchain download (i.e. something running Go `1.21`, since
before then the directive was only advisory, and with the `GOTOOLCHAIN`
env var not set to `none`), for example:

    $ docker run --rm --interactive --volume "$PWD:/work" --workdir /work golang:1.21.0 go mod tidy
    go: downloading go1.22 (linux/amd64)
    go: download go1.22 for linux/amd64: toolchain not available

This change should have no impact on any setup that works with the
current directive.

There's more detail and discussions around this at[4]

[1] https://go.dev/ref/mod#go-mod-file-go
[2] https://go.dev/doc/toolchain#version
[3] https://go.dev/doc/devel/release
[4] https://github.com/golang/go/issues/62278

williammartin reviewed (approved):
Thanks for the detailed issue and the doc references. That Go toolchain doc is quite a read, and not in a good way. Yikes!

I noticed you chose `1.22.0`, is there any reason you didn't choose the latest `1.22.2`? Merging this seems to be fine as is to fix the toolchain issue, but I guess it seems like we should keep moving forward to ensure we have all the security fixes?

matthewhughes934 commented:
> I noticed you chose 1.22.0, is there any reason you didn't choose the latest 1.22.2

This is just the minimum version required to _use_ this module, so I picked the first release of Go 1.22 to avoid enforcing non-strictly necessary version constraints on anyone downstream building this repo:

* If you use go >=1.22.0 to build this repo (e.g. 1.22.2) then you'll build with that version
* If you use go <1.22.0 (and >=1.21.0, since Go before then didn't enforce this verisoning), then:
    * If `GOTOOLCHAIN=none` then you'll get an error like `go: go.mod requires go >= 1.22.2 (running go 1.22.0; GOTOOLCHAIN=local`
    * Else, Go will do a [toolchain download](https://go.dev/doc/toolchain#download) and build using that version

So if CI in this repo is running Go 1.22.2 it should still execute any `go` commands using that version

Example of toolchain download failure:

```console
$ go mod init m
go: creating new go.mod: module m
$ go mod edit -go=1.22.2
$ GOTOOLCHAIN=local go1.22.0 mod tidy
go: go.mod requires go >= 1.22.2 (running go 1.22.0; GOTOOLCHAIN=local
```

williammartin commented:
> So if CI in this repo is running Go 1.22.2 it should still execute any go commands using that version

Right, and we use `setup-go`'s `go-version-file` option to avoid updating go versions in many workflows every time we change the version, so CI uses whatever is in the `go.mod`: https://github.com/cli/cli/blob/05b37e70b2f764a8ff496b57dd7756db50502c08/.github/workflows/go.yml#L49

Perhaps that's not the best option anymore. I suppose it depends if there's value in both:
 * Building our releases with the latest go version
 * Not enforcing unnecessary constraints on anyone depending on this module

matthewhughes934 commented:
> Right, and we use `setup-go`'s `go-version-file` 

Good point, I see CI for this job used (https://github.com/cli/cli/actions/runs/8658525646/job/23742625389):

    go version go1.22.0 linux/amd64

Where as for the latest run on `trunk` (https://github.com/cli/cli/actions/runs/8687585718/job/23821375676)

    go version go1.22.2 linux/amd64

Maybe passing `go-version: stable` (https://github.com/actions/setup-go?tab=readme-ov-file#using-stableoldstable-aliases) suits both cases? Though there might be surprises when that suddenly jumps to Go 1.23 automatically. Another alternative is with `check-latest` https://github.com/actions/setup-go?tab=readme-ov-file#check-latest-version but I think then you'd need to maintain a version spec in two locations (`go.mod` and all the workflows)

williammartin commented:
> Though there might be surprises when that suddenly jumps to Go 1.23 automatically

Hmmm, I think I'd rather be in control of our versions and see failures in branches that wouldn't make us block other changes. For example on the release of `go 1.22.0` there were two breaking changes (https://github.com/cli/cli/issues/8668, https://github.com/cli/cli/issues/8700)

> but I think then you'd need to maintain a version spec in two locations (go.mod and all the workflows)

This returns us to updating the version everywhere which we were trying to get away from unfortunately.

There may not be a "correct" choice here and just a series of tradeoffs. Part of me wants to say "we are shipping a binary, let's just keep moving to latest" but maybe that's just being a bit selfish 😅  Maybe we need to get a change into `setup-go` to make their matrix of options even more complicated!

williammartin commented:
@matthewhughes934 what do you think about:

```
go 1.22
toolchain 1.22.2
```

I've seen a [few different examples](https://github.com/search?type=code&q=%22toolchain+go1.22.2%22) like this e.g. [etcd](https://github.com/etcd-io/etcd/blob/a7f5d4b4e4569bd316277ebf1347785e0467c64d/go.mod#L3-L5).

I know at the start you said "The version must be a valid Go version, such as 1.9, 1.14, or 1.21rc1." but I think the issue here is not that `1.22` isn't a valid go version but that absent of other directives/env vars, it gets used as the toolchain version as well, and **that's** not valid.

This seems like it would indicate:
 1. This module expects language features from `1.22`
 2. The main module should be build with `1.22.2`
 3. `setup-go` should use the latest patch version of `1.22`
 
At least with regards to 3 we would expect it to either be latest patch version of `1.22.2` depending on what `setup-go` actually looks at.

matthewhughes934 committed:
squash! Fix `go` directive in `go.mod`

Keep `go` version relaxed, pair with a `toolchain` directive so that:

* We still get the functional fix (e.g. building with
  `GOTOOLCHAIN=local`)
* We still get the `setup-go` action just pulling in the latest version
  of Go 1.22 when available without having to add a version constraint
  there

matthewhughes934 commented:
> @matthewhughes934 what do you think about:
> 
> ```
> go 1.22
> toolchain 1.22.2
> ```

:+1: I think that should work fine, and do as you described: [1cb8edc](https://github.com/cli/cli/pull/8956/commits/1cb8edce390aaadb3754b6968ef555feca2581b5)

> I know at the start you said "The version must be a valid Go version, such as 1.9, 1.14, or 1.21rc1." 
 but I think the issue here is not that 1.22 isn't a valid go version but that absent of other directives/env vars, it gets used as the toolchain version as well, and that's not valid.

Agreed, I think the docs module docs for Go could help clarity this, I might open an issue in golang (if there isn't one already, I know there's already a couple of conversations around toolchains+mod versions)

williammartin commented:
Thanks a lot for helping us figure this out 🙏  I don't find this part of the Go ecosystem clear at all so having someone else come in and help us work through it is really useful.
